src/content/app-architecture/case-study/data-layer.md

```diff
@@ -0,0 +1,340 @@
+  Future<Result<List<BookingSummary>>> getBookingsList() async {...}
+  Future<Result<void>> delete(int id) async {...}
+}
+
+```
+
+:::note Development vs staging environments
+The class in the example above is `BookingRepositoryRemote`, 
+which extends an abstract class called `BookingRepository`. 
+This base class is used to create repositories for different environments. 
+For example, the compass app also has a class called `BookingRepositoryLocal`, 
+which is used for local development.
+
+You can see the differences between the [`BookingRepository` classes on Github][].
+:::
+
+
+The `BookingRepository` takes the `ApiClient` service as an input, 
+which it uses to get and update the raw data from the server. 
+It's important that the service is a private member, 
+so that the UI layer cannot bypass the repository and call a service directly.
+
+Via the `ApiClient` service, 
+the repository can poll for updates to a user's saved bookings that 
+might happen on the server, and make POST requests to delete saved bookings.
+
+The raw data that a repository transforms into application models can come from
+multiple sources via multiple services, 
+and therefore repositories and services have a many-to-many relationship. 
+A service can be used by any number of repositories, 
+and a repository can use more than one service.
+
+<img src='/assets/images/docs/app-architecture/guide/feature-architecture-simplified-Data-highlighted.png'
+alt="A diagram that highlights the data layer components of an application.">
+
+### Domain models
+
+The `BookingRepository` outputs `Booking` and `BookingSummary` objects, 
+which are *domain models*. All repositories output corresponding domain models. 
+These data models differ from API models in that they only contain the data 
+needed by the rest of the app. 
+API models contain raw data that often needs to be filtered or 
+combined or removed to be useful to the app's ViewModels. 
+When the raw data has been refined, its output as domain models.
+
+In the example app, domain models are exposed via 
+return values on methods like `BookingRepository.getBooking`. 
+The `getBooking` method is responsible for getting the raw data from 
+the `ApiClient` service, and transforming it into a `Booking` object. 
+It does this by combining data from multiple service endpoints.
+
+```dart
+// app/lib/data/repositories/booking/booking_repository_remote.dart
+
+// method edited for brevity
+ Future<Result<Booking>> getBooking(int id) async {
+    try {
+      // Get booking by ID from server
+      final resultBooking = await _apiClient.getBooking(id);
+      if (resultBooking is Error<BookingApiModel>) {
+        return Result.error(resultBooking.error);
+      }
+      final booking = resultBooking.asOk.value;
+
+      final destination = _apiClient.getDestination(booking.destinationRef);
+      final activities = _apiClient.getActivitiesForBooking(booking.activitiesRef);
+
+      [!return Result.ok(!]
```

Comment:
In cases where it's ok to highlight whole lines, consider using the `highlightLines=startLine-endLine` mechanism.

````markdown
```dart highlightLines=2
void main() {
  print('Hello');
}
```
````